smart-contracts/contracts/PublicLock.sol

```diff
@@ -325,6 +325,41 @@ contract PublicLock is ILockCore, ERC165, IERC721, IERC721Receiver, Ownable {
     return owners.length;
   }
 
+ /**
+  * A function which returns a subset of the keys for this Lock as an array
+  * @param _page the page of key owners requested when faceted by page size
+  * @param _pageSize the number of Key Owners requested per page
+  */
+  function getOwnersByPage(uint _page, uint _pageSize)
+    public
+    view
+    returns (address[])
+  {
+    require(outstandingKeys() > 0, "No keys to retrieve");
+    uint _startIndex = _page * _pageSize;
+    require(_startIndex >= 0 && _startIndex < outstandingKeys(), "Index must be in-bounds");
+    uint endOfPageIndex;
+
+    if (_startIndex + _pageSize > owners.length) {
+      endOfPageIndex = owners.length;
+      _pageSize =  owners.length - _startIndex;
+    } else {
+      endOfPageIndex = (_startIndex + _pageSize);
+    }
+
+    // new temp in-memory array to hold pageSize number of requested owners:
+    address[] memory ownersByPage = new address[](_pageSize);
```

Comment:
For me, when first working on `getKeysByPage()` we were working with a dynamic array field on the key struct as well, which complicated and limited what tye of arrays I could use without the compiler yelling at me. With the move from `bytes data;` to `bytes32 dataHash` this may no longer be the case.

Reply:
This is interesting, I hadn't deeply considered the cost of gas. It was more around a reuse scenario. While working with this I did realize that setting the length of the array dynamically avoids callers having to check result for padded values. 

@nfurfaro I will be definitely looking to your guidance as to which way is the best way to proceed here

Reply:
Well, the UI/UX and reusability is better the way you've done it, with dynamic-sizing. Sometimes we just have to accept the gas cost as a tradeoff for useability. I think that given the fact we're reading from memory (not storage) here, the additional gas costs will likely be even smaller. The usability benefits of the dynamic array probably outweigh the additional gas used.

 The 200 extra gas per read above is for storage arrays, and everything in storage is far more expensive than it's memory counterpart (read from storage (SLOAD) = 200gas, read from memory(MLOAD) = 3 gas)

Reply:
@nfurfaro Can you expand on the gas costs? The thing that I do not fully grasp is that this is not a "transaction" (not changing the state) and this will not be called by an identified "sender" so no one will pay gas. 
Now I understand that if we (or someone else) invoke that function as part of an actual transaction, then, yes there would be gas costs... but that's probably not our use case?

Reply:
@julien51 you're correct, I was mistaken. The "view" keyword should have given it away;)